### PR TITLE
refactor: event account refactor v1

### PR DESCRIPTION
**Pull request by obi-ewan**
PLS READ: These are only the necessary fields to replicate on-chain what we currently use off-chain via flex, there are further epics to implement scoring, oracles and settlement

I've added a new `Event` struct to capture the fields we need for v1, the original struct can still be found here https://github.com/BetDexLabs/protocol-event/blob/develop-ci/programs/protocol_event/src/state/event_account.rs#L4

I will remove/update the old struct once we have agreement on v1 fields.

Reading Material:
- Flex event; https://github.com/BetDexLabs/flex/blob/main/src/types/event_types.ts#L47
- Events group JSON for mainnet; https://betdex-data.s3.eu-west-2.amazonaws.com/eventgroups-production.json

### CODE REVIEW

File: programs/protocol_event/src/state/event.rs [hunk 1]
```diff
@@ -0,0 +1,36 @@
+use anchor_lang::prelude::*;
+
+#[account]
+pub struct Event {
+    pub authority: Pubkey,
+    pub payer: Pubkey,
+
+    pub slug: String, // event identifier e.g. LAFCvLAG@2021-08-28
```
Comment (obi-ewan) [inline programs/protocol_event/src/state/event.rs:8]: Not 100% happy with `slug`naming  here as we don't use that anywhere else, but it is in the doc proposal and other APIs (e.g. smarkets) use it

**Review by eoin-betdex — COMMENTED** [review]
Not to clog up the diff with comments. Probably all things you are aware of and I'm just jumping ahead. 

* Might need to think about field ordering if we are to allow for getProgramAccounts filtering by, e.g., `event_group`. Strings unless we do some fixed size padding can't be before anything we intend to filter by. 
* The original consideration for categorisation had more levels (sub-cat and class) but I think these two are fine to start with. We'll probably need more levels for proper organisation of data eventually.
* We'll need a cap on the size of `participants` as well as the different strings so we can size the account appropriately.  
* The placeholder event type doesn't need to be `AVB`, though tbh it can be anything, so long as we only create one type of events for now ;)

File: programs/protocol_event/src/state/event.rs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    pub event_group: EventGroup,
+
+    pub active: bool,
+
+    pub slug: String, // event identifier e.g. LAFCvLAG@2021-08-28
+    pub name: String, // for display purposes e.g. Los Angeles Football Club vs. LA Galaxy
+
+    pub participants: Vec<Participant>,
+
+    pub expected_start_timestamp: i64,
```
Comment (druspencer) [inline programs/protocol_event/src/state/event.rs:19]: Where did we land on `expected_end_timestamp` ? That would be used for figuring out the initial lock time of an inPlay market. Or do we think an operator should calculate that from the expected start timestamp?

File: programs/protocol_event/src/state/event.rs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+pub struct EventGroup {
+    pub id: String, // event group identifier e.g. MLS
+    pub name: String, // for display purposes e.g. Major League Soccer
+}
+
+pub struct Participant {
+    pub id: String, // participant identifier e.g. LAFC
+    pub name: String, // for display purposes e.g. Los Angeles Football Club
+}
```
Comment (pawel-betdex) [inline programs/protocol_event/src/state/event.rs]: EOF

**Comment by druspencer** [comment]
>Might need to think about field ordering if we are to allow for getProgramAccounts filtering by, e.g., event_group. Strings unless we do some fixed size padding can't be before anything we intend to filter by.

fixed string padding would be super useful

File: programs/protocol_event/src/state/event.rs (hunk 2 above)
Comment (druspencer) [reply programs/protocol_event/src/state/event.rs:19]: Operators prerogative based on expected start time ✅

File: programs/protocol_event/src/state/event.rs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    pub actual_start_timestamp: Option<i64>,
+    pub actual_end_timestamp: Option<i64>,
+}
+
+pub struct Category {
+    pub id: String, // category identifier e.g. FOOTBALL
+    pub name: String, // for display purposes e.g Football
+}
+
+pub struct EventGroup {
```
Comment (eoin-betdex) [inline programs/protocol_event/src/state/event.rs:58]: Is `Event` in `EventGroup` redundant? I think it probably makes sense here but just asking the question as I'm aware of all the places we have put redundant prefixes in the past that now are irritating.

Comment (obi-ewan) [reply programs/protocol_event/src/state/event.rs:58]: Yeah good point, no reason to have `Category` and then `EventGroup`, I will remove `Event` for now at least for consistency and see how it looks further down the line

Comment (eoin-betdex) [reply programs/protocol_event/src/state/event.rs:58]: I'm not pushing either way, though I'll be pointing out anywhere we introduce potentially redundant prefixes going forward so we definitely think about them. We did have it on the original hierarchy proposal
```
Category (e.g., Sports, Politics, Entertainment)
\_Sub Category (e.g., Basketball)
   \_Class? (e.g., European Basketball or Women's Baskball)
      \_Event Group (e.g., NBA, WNBA, NCAAB)
        \_Event (Lakers v Bulls June 12 2023)
```

Comment (obi-ewan) [reply programs/protocol_event/src/state/event.rs:58]: If we want to store them in their own accounts further down the line it maybe does make sense to use `EventGroup` as it's a bit less ambiguous than `Group`, I do like `EventGroup` better here but I will have a ponder on it

**Review by pawel-betdex — DISMISSED** [review]

File: programs/protocol_event/src/state/event.rs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+    pub expected_start_timestamp: i64,
+    pub actual_start_timestamp: Option<i64>,
+    pub actual_end_timestamp: Option<i64>,
+}
+
+impl Event {
+    const MAX_EVENT_SLUG_LENGTH: usize = 25;
+    const MAX_EVENT_NAME_LENGTH: usize = 50;
+    const MAX_PARTICIPANTS: usize = 64;
```
Comment (obi-ewan) [inline programs/protocol_event/src/state/event.rs]: Need to think about sizing this, for most markets 64 participants will be fine - there are 48 teams in the next world cup. However, there are 128 participants in Wimbledon events if we were doing an outright; https://en.wikipedia.org/wiki/Wimbledon_Championships#Events

But we actually hit the account size limit when using 128 (`Instruction 0: Failed to reallocate account data`)
 => 140 bytes per participant = 17920 (140 * 128)
 => 0.12 sol for participants alone

Strings are expensive!

Comment (obi-ewan) [reply programs/protocol_event/src/state/event.rs]: update: this won't work for golf (300 players?) need to rethink a solution for participants

Comment (eoin-betdex) [reply programs/protocol_event/src/state/event.rs]: Is this error during account initialisation? There is an initial account limit, but they can be further grown with later additional reallocation instructions.

Comment (eoin-betdex) [reply programs/protocol_event/src/state/event.rs]: Looks like this is a PDA size limit of 10KiBs. Too much data!

Comment (obi-ewan) [reply programs/protocol_event/src/state/event.rs]: We had a good chat about this and I'm going to write up a new proposal for storing event participants, I will update this PR with a placeholder field for now

File: programs/protocol_event/src/state/event.rs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+    pub category: Category,
+    pub event_group: EventGroup,
+
+    pub active: bool,
+
+    pub slug: String, // event identifier e.g. LAFCvLAG@2021-08-28
+    pub name: String, // for display purposes e.g. Los Angeles Football Club vs. LA Galaxy
+
+    pub participants: Vec<u16>,
```
Comment (obi-ewan) [inline programs/protocol_event/src/state/event.rs:17]: This is a placeholder type, Participants proposal still to come!

File: programs/protocol_event/src/lib.rs [hunk 7]
```diff
@@ -5,13 +5,9 @@ pub mod state;
...
-use crate::state::event_account::{Event, EventStatus, EventType};
+use crate::state::event::*;
+use crate::instructions::CreateEventInfo;
 
-#[cfg(feature = "stable")]
-declare_id!("5qCutonYoeg1aRK31mv4oQYoKdNFMpPaEtDe9nnNQXXf");
-#[cfg(feature = "dev")]
-declare_id!("5qCutonYoeg1aRK31mv4oQYoKdNFMpPaEtDe9nnNQXXf");
-#[cfg(not(any(feature = "stable", feature = "dev")))]
 declare_id!("5qCutonYoeg1aRK31mv4oQYoKdNFMpPaEtDe9nnNQXXf");
```
Comment (obi-ewan) [inline programs/protocol_event/src/lib.rs:11]: I've removed the extra program id configs for now, as we didn't need them for the product program. Will add them in as/when required

**Comment by obi-ewan** [comment]
Pls review this PR https://github.com/MonacoProtocol/protocol-event/pull/2

Pull requests are tied to the repository that they were created in, so I can't target this PR at the Monaco Protocol repo

**Review by pawel-betdex — APPROVED** [review]

**Comment by obi-ewan** [comment]
closing this PR as it has been merged up

**Pull request closed without merging**